lightning/restore/restore.go

```diff
@@ -1607,42 +1600,65 @@ outside:
 
 		// sql -> kv
 		lastRow := cr.parser.LastRow()
-		kvs, err := kvEncoder.Encode(logTask.Logger, lastRow.Row, lastRow.RowID, cr.chunk.ColumnPermutation)
+		kvs, encodeErr := kvEncoder.Encode(logger, lastRow.Row, lastRow.RowID, cr.chunk.ColumnPermutation)
 		encodeDur := time.Since(start)
 		encodeTotalDur += encodeDur
 		metric.RowEncodeSecondsHistogram.Observe(encodeDur.Seconds())
 
-		if err != nil {
+		if encodeErr != nil {
 			// error is already logged inside kvEncoder.Encode(), just propagate up directly.
-			return err
+			err = encodeErr
+			return
 		}
 
 		deliverKvStart := time.Now()
-		select {
-		case kvsCh <- deliveredKVs{kvs: kvs, offset: newOffset, rowID: rowID}:
-			break
-		case <-ctx.Done():
-			return ctx.Err()
-		case deliverResult := <-deliverCompleteCh:
-			if deliverResult.err == nil {
-				panic("unexpected: deliverCompleteCh prematurely fulfilled with no error")
-			}
-			return errors.Trace(deliverResult.err)
+		if err = send(deliveredKVs{kvs: kvs, offset: newOffset, rowID: rowID}); err != nil {
+			return
 		}
 		metric.RowKVDeliverSecondsHistogram.Observe(time.Since(deliverKvStart).Seconds())
 	}
 
 	lastOffset, lastRowID := cr.parser.Pos()
-	select {
-	case kvsCh <- deliveredKVs{kvs: nil, offset: lastOffset, rowID: lastRowID}:
-		break
-	case <-ctx.Done():
-		return ctx.Err()
-	case deliverResult := <-deliverCompleteCh:
-		if deliverResult.err == nil {
-			panic("unexpected: deliverCompleteCh prematurely fulfilled with no error")
-		}
-		return errors.Trace(deliverResult.err)
+	err = send(deliveredKVs{kvs: nil, offset: lastOffset, rowID: lastRowID})
+	return
```

Comment:
I think it's better to move L1621-L1622 to above L1590 is better and replace `break` with `return`

Reply:
No, line 1621-1623 will be hit by Line 1575 as well.